### ISSUE TITLE
Airsonic 11.0.0 Release Plan

### ISSUE THREAD

**Issue by fxthomas** [airsonic/airsonic#1439]
This issue is a meta issue for discussing the Airsonic 11.0.0 release. I'll keep this up-to-date as we add more information to the discussion.

This won't change how small PRs and issues are merged, but with this I hope we can be a little more open about how the releases are done. These meta issues were discussed in #1436.

Things to do:

- [ ] Important features / fixes (https://github.com/airsonic/airsonic/milestone/3)
  - [ ] _To be determined_
- [ ] Prepare release
  - [ ] Follow the [release documentation](https://github.com/airsonic/airsonic/blob/master/contrib/release.md)
- [ ] Publish release
  - [ ] Publish release on GitHub
  - [ ] Push release on Docker Hub
  - [ ] Write post on Reddit
- [ ] Post-release
  - [ ] Write meta issue for next release

**Comment by fxthomas** [comment]
Open for discussion : should we do the big Spring update at the same time?

I think a good proof of concept was already done, we just need to break it into minimal changes so that each can be discussed independently.

**Comment by eharris** [comment]
I corrected the PR number for the transcoding changes.  Moved the old PR to the next line in case it was intended to be added too, please correct if not.

**Comment by fxthomas** [comment]
Ah, you're right, thanks! One copy-paste in the wrong place :wink:

**Comment by tesshucom** [comment]
> Open for discussion : should we do the big Spring update at the same time?

The previous PR of the Spring update by randomnicode was split into several components, the initial phase of which consisted of refactoring and minor fixes to related features.

Can't merge the early part to 11.0.0 or 10.6.0?

 - Contains fixes that may help resolve #1340
 - Provide solutions to people suffering from #1340 and make it easier for them to participate in verification cooperation
 - HSQLDB 2.5.0 Update is beneficial. But people who benefit from this are different from users who suffer from #1340. If HSQLDB is prioritized and Spring is all delayed, handling #1340 will be very delayed.
 - Split deployment can reduce the burden and risk of Spring update verification.

Alternatively, extract from randomnicode PR and make a pinpoint correction to #1340.
Slowing down doesn't make a good impression for user.

**Comment by fxthomas** [comment]
Actually, good news : I'm mostly done testing a minimal version of the Spring update, that went faster than I initially anticipated. I think this can be safely added to 11.0.

#1340 seems like a tough nut to crack though. Hikari is worth trying, but I'm not convinced the thread-pool is the bottleneck (otherwise I would have seen a similar behavior with HSQLDB 2.5.0 on my own instance).

**Comment by tesshucom** [comment]
Sounds good.
#1340 seems to need to be investigated separately from this issue.

**Comment by jooola** [comment]
@fxthomas I've been playing around with Github Actions recently, I could set this up in the next days.

**Comment by fxthomas** [comment]
Sounds like a good idea, if we're going to depend on external services, might as well depend on as few as possible, so putting Travis out of the equation has my vote.

I'm not familiar with the service, so when you have results I'd like to hear your thoughts on it! (maybe create a separate issue to discuss?)

**Comment by tesshucom** [comment]
Wm...
This is not an urgent requirement, but I think one of the improvements in the build flow is the independence of integration testing.
Of course, everything should not be done at the same time, so I think it's another discussion in the future.

It's not uncommon for the main project build to be separate from the integrated heavyweight testing.
It increases the portability of the main project and allows the technical implementation of integration testing to be managed separately from the main project.

I do not believe that integration tests should be removed.
However, I feel that there are people who have more happy ideas.

**Comment by tesshucom** [comment]
Thanks to jooola for his visualization work.
Even if we can't start right away, it's very useful.

@randomnicode 
It's a small imagination, but can randomnicode create such a set of suggestions?
In table format and subissues format.
It can be an Airsonic repository or yours.
The purpose is close to publishing a rough estimate of the item and quantity, and to facilitate tracking.
It is unlikely that all of them will be reflected immediately, but it can attract the interest of users and developers.
(I'll tell you, you're a special person, so the average person needs it to get the big picture.)

I think they can be very helpful in building a schedule for 11.0.0 and later.
Also, when dealing with a large number of important matters with practically limited resources, it is necessary to set realistic priorities.

At the forums, some engineers seemed to want to share, to want to ask questions and deepen their understanding of specific issues.
I would rather think it should be done in a development issue, as the FAQ will be a propellant. And the more people that get into the topic, the better.

The problem is that randomnicode has some burden.
So if randomnicode agrees, then.
And we are not in a hurry.
The consensus of the world is that the fight against coronavirus is more important.

I have an idea about searching, meta-processing and DLNA, and it seems like this will be possible in a while.
Judging from the content, the priority of the randomnicode agenda is high.
Depending on the agenda, it may be linked to a jooola issue.

**Comment by tesshucom** [comment]
Thanks to jooola for his visualization work.
Even if we can't start right away, it's very useful.

@randomnicode 
It's a small imagination, but can randomnicode create such a set of suggestions?
In table format and subissues format.
It can be an Airsonic repository or yours.
The purpose is close to publishing a rough estimate of the item and quantity, and to facilitate tracking.
It is unlikely that all of them will be reflected immediately, but it can attract the interest of users and developers.
(I'll tell you, you're a special person, so the average person needs it to get the big picture.)

I think they can be very helpful in building a schedule for 11.0.0 and later.
Also, when dealing with a large number of important matters with practically limited resources, it is necessary to set realistic priorities.

At the forums, some engineers seemed to want to share, to want to ask questions and deepen their understanding of specific issues.
I would rather think it should be done in a development issue, as the FAQ will be a propellant. And the more people that get into the topic, the better.

The problem is that randomnicode has some burden.
So if randomnicode agrees, then.
And we are not in a hurry.
The consensus of the world is that the fight against coronavirus is more important.

I have an idea about searching, meta-processing and DLNA, and it seems like this will be possible in a while.
Judging from the content, the priority of the randomnicode agenda is high.
Depending on the agenda, it may be linked to a jooola issue.

#1538 is another popular proposal.

**Comment by randomnicode** [comment]
Hello, I'm just getting around to seeing this.

Sorry, unfortunately, I can't commit to creating a list of issues. I also didn't understand the part about my agenda priority being high. I think everyone's agenda should bear equal consideration.

My cursory suggestion as a guideline is actually that the scope here for this release is too narrow.

It's a major version release, and you don't know when you're going to do the next one, so you should use this opportunity to do a lot of the major breaking changes you'd like to do. As many as you can. If you can break things, now is the time to break it. Slowness now will cost you later because you're then locked into a version and unless you do another major version again soon (which I don't see happening), you won't be able to maneuver with breaking or "significant changes".

That doesn't just include major feature additions. It also means removing a whole lot of cruft. I see a ridiculous number of unused tables and columns that a major release would be a perfect time to rid of (the old db will already have a backup created due to the HSQL migration). Drop them. Drop all of them. Leave no quarter, hear no mercy. If it doesn't pertain to the codebase as of 11.0, it should be (tearfully if necessary) obliterated. There should be a whole lot of drop tabe x, drop column y, drop index a, drop constraint b migrations.

Similarly, different architectures that require table realignments should be done at this point in time too. A lot of times I see code bending around a restriction instead of just fixing it (because then it won't be backwards compatible). Use this opportunity to just fix the table.

SQL rewrites can and should also be done, especially after HSQL merge given the new DB is now a bit more modern and will likely support more modern features like CTEs. For instance it always grates me how insertion/update logic is performed currently in airsonic, which requires 2 db trips (try update, check rows updated, if 0, then try insert). You're looking for a DB-agnostic UPSERT syntax, which even if it doesn't exist on your DB of choice, you can always emulate (with a slight race condition) with a CTE.

Additionally another annoyance is the whole bit about ids in the db. I often see us checking for a max id and then inserting a row with an id above that, or using the max id as a substitute to get the id of the row it just inserted (which might not be correct if two separate threads added stuff simultaneously). These kinds of issues just prevent airsonic from being scalable in general, less parallelizable, and mask some serious issues.

I'd also get rid of ids in general. If it were up to me, everyone should be using uuids. The client (inserter/updater) provides the id it wants to insert/update, instead of the db keeping track. The only issue is uuid generation for migrations, which I'm sure we can write a function for depending on the db (like postgres needs the ossp-uuid extension, and barring that you can just use a custom procedure to add a uuid generator).

There's a lot of other sql tweaks that should be investigated and done, like IN statements (which I suspect is the cause for a few postgres slowdowns).

Furthermore, the transcoding changes that @eharris suggests are good but not nearly enough. They won't solve issues with different kinds of encodings living inside the same container (which, without actually even looking or doing any kind of investigation, I suspect is the issue behind ALAC playback issues: aac and alac both share the same extension - m4a - and ffmpeg probably defaults to using the aac encoder/decoder instead of the alac, and you can't specify one because you don't input that in your transcoding engine, or even store it).

And of course a hundred million Java related updates.

These are growing pains of course, and painful to navigate, but they're necessary to get a stronger standing on modern tech stacks. For instance, I don't anticipate the intense amount of sql changes pleasureful or even likeable in the very least (and I suspect there's already going to be a pushback at the suggestion), but at the end of the day, you'll have a stronger basis to build and improve on.

I can add new features to these suggestions, but I really don't think you'd want more on the plate. Besides it's easier to add than subtract. So cleaning should be the higher priority.

**Comment by tesshucom** [comment]
> I also didn't understand the part about my agenda priority being high.

It's not a suggestion to treat everyone's agenda unfairly. 

This is an objective observation to date, but in the past there was a large commitment and its reorganization could have slowed everything down. For large changes or questionable updates, issues may require prior consultation. Imbalanced procedures have bad consequences. Without questioning the process, stating only the bad results masks the problem. It is constructive to shed light on bottlenecks and eliminate them, rather than transferring responsibility solely to the development team.

> My cursory suggestion as a guideline is actually that the scope here for this release is too narrow.

If work is pre-scheduled and there is a certain quality of PR, more can be included. Actually, there are many undecided tasks, so I will not dispute the decision of the person in charge. If you have any suggestions, it is best to raise them in issues.


> SQL 

I agree personaly with the need for SQL improvements. As long as it can be run on any DB and prove to be working properly by the tests. Isn't it just that there was no such PR until now? You can raise a topic and convince in advance if someone disagrees with the direction.

If you wish to make a drastic change in design policy, you should see the engineer's reaction in advance in issues. If you send only the PR directly to the reviewer without opinion adjustment, it will definitely stagnate. It's not a reviewer's fault, not mean neglect of reviews but a lack of coordination of the procedure.

I think everyone knows some of the causes of the slowdown. After all, isn't there a need for a consensus on breaking backward compatibility? It's before the technical debate.

___

From here on, my opinion.

Regarding speed-up, while admitting the effect, not top priority.
In the proposition of server acceleration, code changes are generally the last resort.

If we look at the differences in the code only, we run the risk of falling into the illusion of a major improvement.
If it's a risky compared to the benefits, this is a big pitfall.
Because it doesn't match cost performance.

This is the general theory.
Do you want to destroy it?
It's good to have a strategy to kill someone who might object to you.

If you do not want to enumerate the agenda, suggest from the most effective  issues first and make goals clear please.
You should be able to prioritize the actual effects.
Bottom-up proposals are less empathetic, so show your goals and break the process into blocks.
I want proper triage.

In your personal repository, you've probably made changes for each technical topic.
If you want to reproduce that procedure on Airsonic, consider it difficult to convince others. There is room for reconsideration.

Make something that moves correctly, speed it up, and measure the effect.
Normal development is done that way.
Technical discussions follow them, not the opposite.

For example, set a goal of 20% faster scans.
The user will shed tears on the announcement that "scans are 20% faster!".
Set that as a goal.

Then categorize the required fixes top-down.
Isn't it easy to get the consent from the explanation that ""this modification i'll be added for that purpose"" ?
If we have a lot of fix topics, it's easy to understand the plan that "solving in this order is effective".
I would like to have such a story.

Don't assume you can stir your code on a large scale because of vague "speed improvements".
The priority is Scan > REST > Things that only benefit the WEB/HTML5.
It's not my preference, it's a backcalculation of the server use case.

Is it really necessary to break backwards compatibility?
This verification is an extension of those.

**Comment by randomnicode** [comment]
> For large changes or questionable updates, issues may require prior consultation

I disagree. Open source projects do not need prior consultations from anyone. I'm also not sure what a "questionable" change is and who the arbiter is. A "questionable" change to one is a "necessary" change to another.

> Imbalanced procedures have bad consequences. 

Sure, I agree.

> Without questioning the process, stating only the bad results masks the problem. It is constructive to shed light on bottlenecks and eliminate them, rather than transferring responsibility solely to the development team.

I'm not sure what this is referring to. Who is the dev team? Who is transferring which responsibility?

> If work is pre-scheduled and there is a certain quality of PR, more can be included. Actually, there are many undecided tasks, so I will not dispute the decision of the person in charge.

I think that's my point. More _should_ be included. The agenda looks here to be pretty slim. Just a DB upgrade and a transcoding improvement when a lot _can_ be done.

>  If you have any suggestions, it is best to raise them in issues.

I did and continue to do so. Some of my issues turn stale, some get brought up later again as a rehash.

> I agree personaly with the need for SQL improvements. As long as it can be run on any DB and prove to be working properly by the tests. Isn't it just that there was no such PR until now? You can raise a topic and convince in advance if someone disagrees with the direction.

Then author a PR for testing too? Corollary PRs that support a particular design decision or feature don't have to be mentioned. Moreover these topics (particularly regarding testing) have been raised before and a sample PR to follow posted, so it's not quite that "there was no such PR". Please see https://github.com/airsonic/airsonic/issues/1563

Testing is certainly important but shouldn't be a barrier to fix things that need fixing. It's not a "only-if-this-then-that". It's not a trade-off or mutual exclusion.

> If you wish to make a drastic change in design policy, you should see the engineer's reaction in advance in issues. If you send only the PR directly to the reviewer without opinion adjustment, it will definitely stagnate. It's not a reviewer's fault, not mean neglect of reviews but a lack of coordination of the procedure.

By engineer here, I assume you mean the maintainer of the repo? Yeah, I disagree. None of this is the responsibility of a PR author. Drumming up support or whatever. The maintainers can certainly highlight places where they'd like help or wherever from outside devs, and it's good if it comes around, but a PR author is and should be free to do whatever they want and pursue whatever avenue they'd like. Following just a maintainer's agenda just means that it's a maintainer-driven fork, not a "community-driven" fork. An issue raised is nice and for those who don't have time or expertise or willingness to open a PR. A PR is even better and I don't think one needs to raise and issue first before authoring PRs. PRs by themselves are issues.

> Regarding speed-up, while admitting the effect, not top priority.
In the proposition of server acceleration, code changes are generally the last resort.

What server acceleration? Without full context, it's hard for me to to agree or disagree here. All I can say is, per my understanding here, you're wrong. Code changes aren't a last resort. What's your alternative? Tell people to get a beefier machine?

>If we look at the differences in the code only, we run the risk of falling into the illusion of a major improvement.
If it's a risky compared to the benefits, this is a big pitfall.
Because it doesn't match cost performance.

Right, I assume you're talking about some kind of cost-benefit analysis to making changes. I don't think that needs to be said. I think everyone does that regardless when making any change. I'm not talking about unnecessary micro-optimizations.

> This is the general theory.
Do you want to destroy it?
It's good to have a strategy to kill someone who might object to you.

Sorry, I truly have no idea what this means. I assume it's some sort of translation disconnect.

>If you do not want to enumerate the agenda, suggest from the most effective issues first and make goals clear please.
You should be able to prioritize the actual effects.
Bottom-up proposals are less empathetic, so show your goals and break the process into blocks.
I want proper triage.
> Make something that moves correctly, speed it up, and measure the effect.
Normal development is done that way.
Technical discussions follow them, not the opposite.
>For example, set a goal of 20% faster scans.
The user will shed tears on the announcement that "scans are 20% faster!".
Set that as a goal.

No.
I'm not running sprints for you guys.
And please don't tell me how "normal development" is done. Without addressing the specifics of what you think "normal development" looks like, I would know, I think.

> Don't assume you can stir your code on a large scale because of vague "speed improvements".

Nothing I've proposed in this repo or the advanced fork has ever been "vague" by any means, so I'm not really sure what you're talking about.

> The priority is Scan > REST > Things that only benefit the WEB/HTML5.

I don't know where this priority or agenda comes from, but sure. That's your priority. You should add these items to your list. I'm not even sure what REST here means, but I assume it means something to you. Similarly, I don't know what Scan is here (like the media scan?), but again, sure. That's not a big architectural change however. It's a tiny PR focusing on the media scan. I remember posting it here ages ago to parallelize it. Just parallelize it and you'll see benefits. The html changes are actually pretty important too. The page rendering is awful (and not HTML5).

> Is it really necessary to break backwards compatibility?
This verification is an extension of those.

You must be joking. You're incrementing a major version and asking that question?
If nothing else, the first item on this topic's agenda should tell you that you're going to break back-compat. The whole topic is about _11.0_ and you're still debating whether breaking back-compat is necessary?

**Comment by randomnicode** [comment]
> Please, there is no need to be overly aggressive about this

Me? Not trying to be. I'm just trying to understand why that conversation was necessary. I was asked for input, I said I can't make lists, but here's a loose rubric of what my suggestions would be. Instead (it seems to me) I was told about "normal development" processes and questioning whether breaking compat is necessary. Maybe it's my difficulty in understanding translated text.

> Well, nothing's been really started yet.

Right but I thought this was the agenda for the next release? Doesn't have to be started.
Why don't we add everything, and take it out later if it doesn't fit some arbitrary time constraint (which I don't see any of)?

> Get the big upgrades (Spring/HSQLDB/JDK11/Docker integration tests) done

Again, these, to me, aren't big upgrades. These are relatively minor upgrades. Spring shouldn't even merit a major version increment. Increasing a major revision of a dependency, doesn't to me count as a "big" upgrade by itself. A big upgrade, in my opinion, is something that architecturally or fundamentally changes something about your software, and that's what I'm urging to do. All these are just minor tweaks to base dependencies. You stored data in a relational db prior, you're still storing it in relational db. Nothing has changed in that regard. You used a JDK prior, you're still using a JDK. And so on. This is the time for "big" upgrades. Not just minorly incrementing some base versions. Incrementing those revisions gives you a base and ability to actually do those big things (like changing the underlying sql because the new relational db can now support modern syntax or data types).

> ANTLR search improvements (might be interesting for starting the implementation of Smart Playlists)

Lucene?

> Use of datatables in the UI (same thing, might be useful for Smart Playlists)

I think you'll need a little more behind this. airsonic uses MVC to generate server-side html which makes it difficult (or useless) to do client-side stuff like data tables. You'll need to start doing stuff the REST or websockets (or combination) way.

> Parallelization of media scans (the performance gains are enough that there's a big benefit in bringing it in Airsonic)

This isn't the case of gripes, but just a pointed note to underline the uselessness of artificial slowdowns in development and how it can actually hinder contribution and advancement of a fork. There was a big benefit in bringing it to Airsonic half a year ago too. These same exact performance gains in the same PR were proposed way back in https://github.com/airsonic/airsonic/pull/1385 and the benefits were exactly the same then as they are now; it's not like they were hidden. For whatever reason it wasn't merged, either the performance wasn't believed, or some other reason. The PR languished for about a month before it was finally taken down. It took almost half a year (and a separate fork) for people to actually realize its potential or worth, and at the end of the day, what gets merged will be something pretty similar to what was proposed a half year earlier. Wouldn't it just have been easier to merge it then so we could've saved half a year's worth of everyone's time? What was the point of the delay or hesitation then? Untrustworthiness of some random online entity trying to make something better? Why not just look at the code and gauge the nefariousness of purpose? Similarly, see the PR for Spring upgrade. Or my HSQLDB upgrade (that is at this point ~3 years old now?). I mean it doesn't just have to be me as an author. That would have been anyone's experience. What did an extra half year grant anyone other than wasted time? The PR is will remain the same, so will the underlying concept. The tech will too. It just took that long to come around to the same conclusion that was reached by someone half a year ago (that was even pointed out as the gain/benefit in the PR).

> Eventually, Sonos/UPnP things we mentioned in another thread (but I can't really test them, so for me it's a second priority and I would rather have somebody else help)

I think this is just a library change. Or decision to take out DLNA. Not a big change, in my opinion. I mean okay to put it on the list, but this is like a side issue that would make it into this release.

I would urge and advise a cleanup. I remember an @eharris PR about refreshdelay seconds. That column needs to be dropped. media_file pulls data from some obsolete table, should drop it. I would delete the user prefs table and rewire them because it is annoying to run a migration every time you want to add a new user pref (like a adding or removing a simple boolean switch whether to enable/disable a third party is a big production). Just put user settings in a json column or something (might have to think about Mysql lengths), it's not like individual columns get queried. Use of CTEs for upserts, change to UUIDs. Switch to using HTML5. etc etc etc.

**Comment by tesshucom** [comment]
There seems to be some misunderstanding, but I have never opposed PR for randomnicode. Rather opposed it being discard before.

Commit size was often an issue, so be careful about PR sizing. The maintainer's past remarks on it are not unreasonable. Summary is important. And,  issue new Issue please instead of discussing cleanup here. Just a little, respect the manageability and visibility of others. 

#1385 looks unsolved, combined with #1450, right?
I think it is necessary to continue if necessary, but how about that? (Don't start here)

___

I can help UPnP, but before that I want a formatter. If possible, it is ideal to be improved by members who have pushed this far, but the final answer would have been to do nothing. If not motivated, either one will do. Airsonic was just reduced outernal support itself.

If you only need ANTLR, you can focus only on the UPnP search function. The range of other UPnP features is wider than you might think. There are many tests. Personally, I thought #1619 would be better.

**Comment by fxthomas** [comment]
I'd write "any and all architectural changes" in the list, but that doesn't tell much if someone doesn't do it... I will add items to the list myself when I have a clearer idea on what I want to add.

> #1385 looks unsolved, combined with #1450

I'm currently testing a successor of #1385 from randomnicode's work, and I think with that we can close both of them. I'll include it in the list.

> I can help UPnP, but before that I want a formatter.

Are you using one in jpsonic? If you can recommend one I am okay with it (I don't know Java formatters, so I prefer to reuse existing work).

> Personally, I thought #1619 would be better.

I'll get it merged :wink: 

----

Off-topic, re:development process.

> And, issue new Issue please instead of discussing cleanup here.

Honestly I feel mailing lists would be easier for these kinds of topics than GitHub issues. Oh well.

> It just took that long to come around to the same conclusion that was reached by someone half a year ago (that was even pointed out as the gain/benefit in the PR).

Here's how I see things : getting multiple people to review, trust and agree on a PR means everyone has the time to understand what's being done and get more familiar with the code, and everyone gets more confident in handling it afterwards.

Of course this kind of process will be much slower than the alternative you describe, but I guess I quite like it in spite of this? I was very unfamiliar with a lot of the code base last year, and I feel this helped me a lot.

Now that I'm a bit more familiar with how things work, I feel more comfortable with actively pushing for inclusion of some PRs, including reviving ones I could only give a passing review earlier.

**Comment by tesshucom** [comment]
> I'd write "any and all architectural changes" in the list, but that doesn't tell much if someone doesn't do it...

We will leave the final grain size to you. As for commits, if it is appropriate, you can use even 1 commit per topic. (Lucene was like that, right? The review process was complicated, but at the end it was squashed.) Sometimes it is the optimal solution.

At the very least, we shouldn't have a mix of Lib update changes, simple cleanups, and bugs. Makes subsequent tracking difficult. This was almost the same too (when updating Lucene). In the middle of the update work, we have fixed the defect of the legacy and rebased. As you probably know, the scan is still flawed. If you are working on it in this fix, split the commit. It can be a stepping stone in resolving unsolved defects. (That's why I'm trying to organize issues. ) To be honest, I want to finish the issue cleanup earlier than the v11.0.0 review.

It's not the responsibility of the committers or reviewers to be prone to simple cleanups and bugs when modifying large Lib updates in old severs. The debt and characteristics of legacy code, and this is one of the situations and common. It's frustrating, but there's no point in being too aggressive.

**Comment by tesshucom** [comment]
Isn't there  the thread residual verification test for MediaScanner ? After all, it's an unpaid job, so I don't insist on the need for exception testing. Although, there are concerns about doing a trial and error level merge.

**Comment by tesshucom** [comment]
I'm not against the idea of ​​parallelization, but it's not a patch or enhancement. It's a change of architect. Usually, after confirming the design change based on the requirement definition, implementation is started.This isn't unique to this project, and I  think I'll be hard to accept implementation reviews.

> Don't assume you can stir your code on a large scale because of vague "speed improvements".

It may not have been communicated to English-speaking people because my writing was not good.
For the fastest progress, please issue issues that interact with requirements and design. At that stage, it's a good idea to get consent from muff1nman and other members. It's a different level than changing JavaScript. It is an architect change that can be affected from outside the service and can affect the outside of the service. From a general point of view, there should be common sense approaches and theories.

When such an agreement is made in advance, the division of tasks after that can be controlled to an appropriate level of granularity and the review perspective is clarified. If not, it causes a great loss.

**Comment by fxthomas** [comment]
@tesshucom I'm trying to understand, do you mean adding an issue like this to organize ideas?

```
We have a lot of performance issues related to media scanning (#1340, #1299).

In order to solve them, the following solution could work but need investigation:
* Parallelization (existing implementation in #1662)
* Data structure changes
* Database query improvements
* ...etc.
```

...and then discuss things on this issue on a higher level? Would that work?

**Comment by fxthomas** [comment]
I added a new issue (#1678) to discuss smart playlists, some of the dependencies could potentially be done for 11.0.

**Comment by tesshucom** [comment]
> do you mean adding an issue like this to organize ideas?

No, we assume a little more upstream point for engineers and engineers to gain a consensus on the design.

In a simple example, the point "is it a legacy replacement?" Single-threaded scans and multi-threaded scans can be implemented in parallel. (Airsonic is not an embedded server for any particular hardware.) Discussion of whether it should be optional should be done before implementation.
In addition, there are thread control methods and exception detection policies. In some ways, it's possible to write terrible programs that leave threads on failure. We should start with the exception design.
In addition, it is possible to present the design in advance which processing blocks are desired to be parallelized. Whether it is appropriate can be verified before implementation. There are some things you may be worried about, such as shutdown?  These agreements require at least muff1nman.
These are not the contents that will be brushed up by trial and error after implementation. If pre-implementer are unable to present some of these views, it implies that are no longer in control already.

The point of view presented by fxthomas is automatically determined if the prior conditions are fixed.(It does not mean that it is not important.) When those tasks are ready to be finalized, it is possible to split the PR into several parts and plan a gradual merge. When updating Lucene, you had some size concerns, right? In fact, the design review and merge were in stages, with all updates that did not impact the existing implementation pre-merged. The migration is bug free.(Although other updates made to others after that have caused interfering bugs.)

Take a deep breath and think. It needs to be run without any bugs. This is because unlike a small patch bug, it directly leads to a server death. This kind of bug will seriously undermine reliability.
It also means that once we change architect, it will be maintained for a long time. (Years later, it may not be the people here who are doing it)
The important thing is not to realize it immediately, but to realize it without mistakes.

**Comment by tesshucom** [comment]
> do you mean adding an issue like this to organize ideas?

No, we assume a little more upstream point for engineers and engineers to gain a consensus on the design.

In a simple example, the point "is it a legacy replacement?" Single-threaded scans and multi-threaded scans can be implemented in parallel. (Airsonic is not an embedded server for any particular hardware.) Discussion of whether it should be optional should be done before implementation.
In addition, there are thread control methods and exception detection policies. In some ways, it's possible to write terrible programs that leave threads on failure. We should start with the exception design.
In addition, it is possible to present the design in advance which processing blocks are desired to be parallelized. Whether it is appropriate can be verified before implementation. There are some things you may be worried about, such as shutdown?  These agreements require at least muff1nman.
These are not the contents that will be brushed up by trial and error after implementation. If pre-implementer are unable to present some of these views, it implies that are no longer in control already.

The point of view presented by fxthomas is automatically determined if the prior conditions are fixed.(It does not mean that it is not important.) When those tasks are ready to be finalized, it is possible to split the PR into several parts and plan a gradual merge. When updating Lucene, you had some size concerns, right? In fact, the design review and merge were in stages, with all updates that did not impact the existing implementation pre-merged. The migration is bug free.(Although other updates made to others after that have caused interfering bugs.)

Take a deep breath and think. It needs to be run without any bugs. This is because unlike a small patch bug, it directly leads to a server death. This kind of bug will seriously undermine reliability.
It also means that once we change architect, it will be maintained for a long time. (Years later, it may not be the people here who are doing it)
The important thing is not to realize it immediately, but to realize it without mistakes.


> Regarding speed-up, while admitting the effect, not top priority.

Stable use by users is a top priority.

**Comment by tesshucom** [comment]
#1470
#1708

Do you think these will be parallelization blockers?
The risk known before is not called growing pains.

**Comment by fxthomas** [comment]
> Do you think these will be parallelization blockers?

At least #1708, yes. If someone has an unambiguous album structure (one album <-> one directory) then everything in #1662 should work fine. But it isn't the case for a lot of users.

For #1470, I think it is not affected by scanning.

**Comment by tesshucom** [comment]
Wm ...

#1385 (wip) has to be judged to be insufficient for review now. Anyone can pull request. But, recognize that we need a complete preservation of legacy, a migration method that is error free, and the proof. No trial and error is allowed. 

The reason I created a redundant PR for Lucene was because there was a PR that was under-validated before that, and stagnation could be expected. The maintainers at the time were making proper decisions and rejecting them. Similar things are about to happen.

Again, I'm not against parallelization. You may not like it, but note that poor design and poor validation are features that should not be admitted. More estimates are needed than non-technical users think. The relevance is not clear, but there is a server dead report at reddit.

I think the brainstorm has progressed more than in the early days. That's a good point. I think it's a good idea to cool down once and do other tasks. It's not a good to be stagnant all under the influence of one PR...

**Comment by randomnicode** [comment]
> #1385 (wip) has to be judged to be insufficient for review now.

Has a successor that is being actively reviewed.

> But, recognize that we need a complete preservation of legacy,

No, we don't. You're doing a major version release. It's in the title. You don't need to keep legacy. Read and understand [semver](https://semver.org) before making these comments.

> a migration method that is error free,

Sure

> and the proof.

I assume this is referring to the PR mentioned earlier. How do you intend to prove? Tests? Both methods pass the same tests. Error rates? All the folks on [`airsonic-advanced`](https://github.com/airsonic-advanced/airsonic-advanced) have been using it without issues for about half a year now.

> No trial and error is allowed.

You don't have a release. Even if you merge something faulty, no one except devs see it because there isn't a release. You don't even have a preview release.

But in any case, this bell has been wrung before, and this point been made many times. Do you need things rock solid for 11.0 release? Sure. But one prerelease PR doesn't matter when a followup can fix things before things go out to public. In fact, you already do those anyway.

> Again, I'm not against parallelization. You may not like it, but note that poor design and poor validation are features that should not be admitted. 

Ahem, sorry, I didn't see you mention any _actual_ technical issues with the design. Did I actually miss some comments? Because I sure would address them if you had raised a question about specifics of the architecture...

And I'm sure you're "validating" by looking at the PR, right? Because that's the point of review... Do you see any issues looking at the code? The change isn't massive and the performance implications are big. "Poor validation" is a function of poor review, which isn't really the PR author's domain. So go and read the PR and raise concerns over there if you have actual concerns or issues or don't understand something or want to learn more about how something is being done that you don't grasp.

> More estimates are needed than non-technical users think.

So do them! Are we trying to preach or lead by example?

> I think the brainstorm has progressed more than in the early days. 

I didn't think there was any brainstorming happening in that PR.

**Comment by tesshucom** [comment]
There seems to be a misunderstanding, but the proof by the test should be done by the submitter. Reviews are not a place for verification, trial and error. Well thought out tests are reflected in the design. The creator of a complete design can explain skillfully the risks and points to others. The cause of the stagnation is the way in which the work of enforcing "proof of the devil" is proceeded.
Of course not all PRs apply to that. Especially when it comes to a certain level of complexity and high risk. We judge that the current deliverables are not well understood and verified, at least as a maintainer.

**Comment by randomnicode** [comment]
> There seems to be a misunderstanding, but the proof by the test should be done by the submitter.

There _are_ tests. And they _are_ passing. See that green tick in the PR?
The media scan in parallel is clearly working, or your tests that rely on a scan wouldn't be able to discover the media files. You'd have a bunch of test failures. There's even a test specifically for a scan. That too passes. Moreover, you already have people in the field for whom it works. And for half a year too without complaints! What other "proof" are you looking for?

> The creator of a complete design can explain skillfully the risks and points to others.

Hmm, did I not? My original PR is literally littered with annotations and comments. Did you read them? I've since answered most of the clarifications posed. Did you read them?

And even beyond that, this is a _PR_, actual _reading_ of the code is expected from a reviewer. This isn't a technical writing 101 class for folks who don't want to read code. Otherwise you shouldn't be a reviewer. There are no design documents to be submitted. Did you read the code?

If you don't understand the code, either ask for clarification, or bow out, instead of being a blocker.
If you haven't read it in full, then you shouldn't be commenting on it because how are you even going to understand what is going on?

> We judge that the current deliverables are not well understood and verified, at least as a maintainer.

So then don't take the PR? I'm not sure what the argument here is? No one is forcing you to?
@fxthomas wanted to take the PR and presumably was asking some clarification about the design itself, and I was trying to understand and answer his questions?
Your comments on the PR seem to be less about the PR and the code itself and more about the vagaries and philosophies of coding practices, none of which, frankly, seems to be constructive or helpful for the PR.

**Comment by randomnicode** [comment]
It seems this issue has seemingly veered off-topic, and we should bring it back and keep this focused on 11.0 plan.

My apologies for partaking in a discussion that was clearly starting to go off topic anyway.

**Comment by tesshucom** [comment]
> And for half a year too without complaints!

What does this guarantee? If yesterday's scan and  scan after 100 years' are the same conditions, does elapsed time have any special meaning? I think it is impossible to design, test and implement unless it is test driven. At the level of finding flaws before reading the code, It is too early in a milestone.
If corrective, it can be persuasive. Even after that, it's not too late. A fix that should be treated more carefully in the sense that it can break things. 
 I think it makes sense to adjust the priorities of release as there are other PRs that are still useful. This does not mean that parallelization is not important. This is a timing adjustment consultation and this is the place to coordinate such plans.

**Comment by randomnicode** [comment]
Sigh

> What does this guarantee? If yesterday's scan and scan after 100 years' are the same conditions, does elapsed time have any special meaning?

Yeah! If a piece of software has been in the field for a 100 years without complaints, you're pretty damn confident that it's robust! I believe the phrase is "tried and true".

> I think it is impossible to design, test and implement unless it is test driven.

Are you just not reading my comments? There ARE tests, and it PASSES them. Do you need me to reiterate?
FURTHERMORE, you have actual real testers in the field for half a year who seem to have been using it without reporting any bugs.

> At the level of finding flaws before reading the code, It is too early in a milestone.
If corrective, it can be persuasive. Even after that, it's not too late. A fix that should be treated more carefully in the sense that it can break things.

After all this verbiage, I'm still waiting for you to tell me precisely where the "flaws" are and what "fixes" there need to be made and what is "breaking things". These words in a vacuum are meaningless. Read the PR and point out the issues (if there are any).

The tests pass. If you want to add more tests, suggest them or add them yourself.

---
> I would delete the user prefs table and rewire them because it is annoying to run a migration every time you want to add a new user pref (like a adding or removing a simple boolean switch whether to enable/disable a third party is a big production). Just put user settings in a json column or something (might have to think about Mysql lengths), it's not like individual columns get queried.

On a different (actually useful) note, this is what I was envisioning for user preferences: https://github.com/airsonic-advanced/airsonic-advanced/pull/283/commits/62b448487992dd926af9748a86e1e1a702fbf9d9

I now effectively can add and remove user prefs without a care in the world, free of any db migrations.

> Here's how I see things : getting multiple people to review, trust and agree on a PR means everyone has the time to understand what's being done and get more familiar with the code, and everyone gets more confident in handling it afterwards.

The issue with this entire philosophy is that while it may work for an individual, it'll never work for a group, because new people are constantly coming in and old people are phasing out. You will always be slow because new folks will take time coming up to speed with the intricacies of a codebase. So _everyone_ actually isn't getting more confident with handling it afterwards (only _you_ are). If you need to search for the truth of this statement, you don't have to look far, you can just look at the existing codebase of airsonic and see it for yourself. New people don't know why certain stuff is there and folks who put it in there have either forgotten or no longer contribute. Take it down another 6 months into the future and this cycle repeats.

I'm not advocating rushed development or anything, but I'm also saying taking multiple years to get a feature in is clearly wrong and a sign that your development processes or philosophies need revision.

>Of course this kind of process will be much slower than the alternative you describe, but I guess I quite like it in spite of this? I was very unfamiliar with a lot of the code base last year, and I feel this helped me a lot.

It _will_ be slower. And while that suits your timeline, think of other folk's time you just wasted (sometimes many months, sometimes years). Just for your own instruction, or unfamiliarity, you slowed down the development of an entire repo (I'm not saying you in specific, I'm pointing to the general case you). Multiply that by everyone who is unfamiliar, and you've got a major slowdown on all fronts.

So your liking or disliking has nothing to do with it. It's a communal project. You're simply its vanguard.

**Comment by tesshucom** [comment]
It's a joint project, but it will be operated by a maintainer who is unpaid, and operated by the maintainer invited by the owner. One-way conversations and hate speech that ignore situations and priorities do not seem to improve the situation. 
Rather.

Ever since humans were able to use fire, no media server project has completely satisfied its users.
So what we have to do now is drink tea.

HSQLDB upgrades (mreged #1345 -> related #874, #1575)

Looking at the existing conversations in Issues, it seems like this relationship (now).
The basis for the judgment looks like this :

> If your server actually crashed (as in, kernel panic/freeze or hard power off) there's a good chance the HSQLDB files got corrupted. The version we use (1.8) is old and apparently not that reliable if the underlying hardware has issues.

There is an analogy that this could be improved.
Therefore, please add the links of #1345, #874 and #1575.

**Comment by randomnicode** [comment]
> One-way conversations and hate speech that ignore situations and priorities do not seem to improve the situation.

Dude, maybe you don't understand the implications, but "hate speech" is a pretty strong and serious accusation, and extremely unwarranted.
You need to seriously stop talking. Address the points on their technical merits or move on.
All you're doing is talking now for the sake of it (and in lieu of an actual technical counterpoint, levying pointless accusations left and right).

**Comment by jooola** [comment]
Could you please move this discussion on the community channels ?

**Comment by randomnicode** [comment]
Agreed.

Or better yet, not have it at all. Waste of everyone's time...

**Comment by tesshucom** [comment]
If we release #1345, you should consider merging #1750. This is because users who have already deployed Airsonic on Tomcat will be in trouble. First, let's run it correctly.

**Comment by tesshucom** [comment]
@randomnicode

Sure, Airsonic may not mention domain and product name diversions, but documents from large communities like Spring and Apache explain what they mean. What you were doing is abuse to maintainers or malicious spoofing as a result. How you look at it. It may not be against the law of your country, but does not comply with some norms or socially recognized patterns of behavior or thinking. The your email also makes me smell it.
Isn't it necessary to make the appropriate brand changes, as muff1nman did? And you can work in a community other than the Airsonic community. It's a better way to do it at your own risk, without attacking others and without fooling users.

**Comment by jooola** [comment]
@lrusak I would say Airsonic is lacking contributor time.

I am not the best person to answer, I haven't contributed here for a long time. Still.

It is an open source project, and it seems the project didn't receive a lot of contributions this past year. I guess most of us are busy doing something else and not having much time.

Some things could be improved and fixed, but the software is working as expected. I mean I have it running for years and I am really happy about it.

As I understand it, @randomnicode was not happy with the pace of this project and wasn't willing to wait for community reviews for his contributions. So he forked. I don't know where his project is heading, but it seem to have some good improvements, haven't tested it.

**Comment by jooola** [comment]
I would be really happy to see Airsonic and Airsonic Advanced to merge together, but I think this will require effort from both sides in order to work. I don't know if you @randomnicode like the idea ?

**Comment by lrusak** [comment]
@jooola no worries. I'm still happy with Airsonic. I'm just a bit disappointed as it seems subsonic in general is a very forked project (madsonic, libresonic, airsonic, etc). I've used many of the forks but in the future I don't want to have to change forks again.

If I can help contribute in some way I would be more than happy to though I don't have much experience in java and javascript.

**Comment by jooola** [comment]
@lrusak Helping out on the bug tracker is already something good. Documentation as well.

**Comment by randomnicode** [comment]
> What you were doing is abuse to maintainers or malicious spoofing as a result. How you look at it. It may not be against the law of your country, but does not comply with some norms or socially recognized patterns of behavior or thinking. The your email also makes me smell it.

@tesshucom I have no idea what you're talking about. I'm not sure what my email makes you smell (which email?). Nor do I understand what you mean by law of country or social norms. What are you talking about? What spoofing?

My contributions to this fork as well as any others stand on their own merit. I engaged in no malicious behavior, tried to plead my case and only after being repeatedly frustrated, very hesitantly decided to fork, which, in retrospect, looking at the progress here, seems like a good decision more and more.

I didn't make anyone stop contributing here. In fact, I warned this outcome if policies weren't changed, to no avail.

> Isn't it necessary to make the appropriate brand changes, as muff1nman did?

No. GPL.

> It's a better way to do it at your own risk, without attacking others and without fooling users.

Huhh? Who have I attacked or fooled?

>  wasn't willing to wait for community reviews for his contributions.

@jooola I waited on some PRs for more than 1.5 years! Others were shut down without a look. I think a more accurate statement is I didn't want to wait _any longer_. 

> I don't know if you @randomnicode like the idea ?

I do, I'd be very much in favor of remerging the two again. I've never wanted a fragmented community. But no idea where the maintainer of this repo is or is interested.

**Comment by tesshucom** [comment]
Take it easy first, then you can do it.
All I'm mentioning is the project name and domain. It's pretty special attempt to include the parent project name completely. This is often forbidden in large projects. It usually changes. Well this is customary as an engineer. As mentioned earlier, there is no rule in Airsonic that prohibits this.... so it's not exorbitant, even if we feel something is wrong or spoofing. But you can avoid it when you do something similar in the future on the other project.

> But no idea where the maintainer of this repo is or is interested. 

Well, I agree with that, but there are ideals and reality. Certainly the reviews were stagnant. I think we need to make an effort to understand why it happened. I read most issues, but at least in the early stages there was evidence that the patch categories to be accepted were controlled.(It is also described in the document) It was called Patches Welcome. There were also a statement of belief that maintainers' resources are limited. These should not be ignored.

Ideally, maintainers should take all the issues you receive from your users seriously. This may be in the user's mind. But ... Perhaps this project was a bit narrower in scope. I guess so from a long time ago document. Therefore, those who have paid so much effort to achieve this project should not be blamed more than necessary. Slow, fast, it's subjective. It may have forced our values on them. 

To put it quality point of view, it's easy to merge high quality code into high quality code. Everyone knows. Otherwise it will be difficult. I feel like we can't talk about what we are aiming for without the owner. However, the distributed repository is a mechanism for others to take over when a stagnation occurs in the central repository. Depending on the circumstances, you will have to accept it.

**Comment by randomnicode** [comment]
> Therefore, careful projects explicitly prohibit duplicate project names. For harmonious management for all subprojects and forks..

Disagree. There are no such "implicit" contracts between devs. The only thing governing software use is the license, which in this case is GPL and all anyone needs to worry about. In fact, most fork names arguably are derivatives of parents or retain some part of the original name to identify where they forked from. Some counterexamples to your "prohibitions/forbidding of duplicate naming in large projects" include Bitcoin (BTC -> Bitcoin Cash or any of the bazillion clones), SSH (SSH -> OSSH -> OpenSSH) etc etc.

As far as categorization and triage goes, you can (a) ask them, or (b) assume it's airsonic unless they mention 'advanced', which is generally what we do.

**Comment by tesshucom** [comment]
I'm not asking for consent. I just asked and you answered. 
Well, your project is a misleading project name, but it's not [part of airsonic](https://github.com/airsonic). Of course there are no special plans for that. It seems that no maintainer already feels the merits of accepting your proposal, for the effort. Anything else?
